### ISSUE TITLE
Code Integrity Check

### ISSUE THREAD

**Issue by MTRichards** [owncloud/core#19036]
As an ownCloud Admin, I want to know if my ownCloud apps and core ownCloud code have been altered so that I can ensure security of the code (hacked, or patched), and better predict conflicts during an upgrade process

Acceptance Criteria:
- An app delivers this capability (enterprise only) and can be enabled or disabled by the admin as any other app
- The app is shipped with ownCloud Server, and supported for Enterprise customers
- App is called “Code Integrity Check”
- A button exists in the admin panel that says “check ownCloud integrity”
- When button is pressed, ownCloud core integrity is checked against the stored signature
     - A badge is shown Green if core is unchanged (Integrity Check Successful.)
     - Badge is shown Red if changed (integrity check failed. ownCloud core has been modified.)
- App also includes a signature of the app itself, so that the integrity check is run first against the integrity check app
- Config file and data directories are excluded from the integrity check
- Individual apps inside ownCloud each store their own integrity checksum / signature. Any app beyond core is checked against this signature when the check integrity button is pressed, or at upgrade. 
- For each app that is installed (whether enabled or disabled), the checksum is checked and a line is listed: App name – with the integrity check results next to id.
     - A badge is shown Green if core is unchanged (Integrity Check Successful.)
     - Badge is shown Red if changed (integrity check failed. ownCloud core has been modified.)
- If an app fails, it gets a red light and the check moves onto the next app.
- Even if the integrity app is disabled, the app is run as part of the upgrade process (unless it has been removed). 
- Before an upgrade, the integrity check is run. If it passes on core, the upgrade is allowed. If it fails on core, the upgrade errors on integrity check “This ownCloud code has been modified from it's original codebase. This can happen for both legitimate and illegitimate reasons, but upgrades with modified code can produce unpredictable results, and even lose data. Please verify before proceeding." Then provide a cancel or continue button. Cancel clearly says upgrade failed.
- If the code integrity fails on upgrade for a specific app, or apps, then throw a warning "Code Integrity Check has failed on the following ownCloud app(s): list. They will be disabled by the upgrade process as upgrading apps with a failed integrity check can produce unpredictable results."
then provide a cancel or continue button and disable the apps in question.

Likely there are more cases here, such as what to do after an upgrade is cancelled, because the system is now in a not working state - with the new code, but the old database. the admin then has to replace the new code with old code for example.

**Comment by RealRancor** [comment]
Hi,

looks partly like this discussion here: https://github.com/owncloud/core/issues/18309

**Comment by LukasReschke** [comment]
> As an ownCloud Admin, I want to know if my ownCloud apps and core ownCloud code have been altered so that I can ensure security of the code (hacked, or patched), and better predict conflicts during an upgrade process

You can't ensure the security of the code that way. Not possible as long as config.php is a PHP file which is writable instead of read-only.

**Comment by LukasReschke** [comment]
> An app delivers this capability 

Please not. This should be part of core. – An app checking the integrity certainly belongs into the core code such as the code checker does etc.

> (enterprise only) and can be enabled or disabled by the admin as any other app

Enterprise only? I'd highly advocate against this.

**Comment by karlitschek** [comment]
Yes. This has to be part of core to work properly.

**Comment by MTRichards** [comment]
Happy to have ideas @LukasReschke - conceptually do you see what we are trying to do? Make sense?

**Comment by LukasReschke** [comment]
> conceptually do you see what we are trying to do? Make sense?

Yep :-)

**Comment by PVince81** [comment]
What happens if there are apps without the checksum ? (ex: from app store)
Does it block the upgrade with the same warning too ?
Or is that only for shipped apps ? (maybe I mixed this up with app signing on the app store)

**Comment by karlitschek** [comment]
The exact consequences have to be discussed. I think a big red warning with the option to override it is the best. Like it's done in Android

**Comment by PVince81** [comment]
PR is here https://github.com/owncloud/core/pull/20285

**Comment by PVince81** [comment]
- [ ] integrity check routines: https://github.com/owncloud/core/pull/20285
- [ ] call it from the upgrade code @VicDeo

**Comment by VicDeo** [comment]
@LukasReschke what is our policy for unsigned apps? 
To be exact for 9.0. And what if integrity is broken - report to user and ask `Do you want to proceed on your peril?`

**Comment by jospoortvliet** [comment]
@VicDeo https://doc.owncloud.org/server/9.0/admin_manual/issues/code_signing.html shows what happens on an installed installation.

TL;DR: In general, unsigned apps for experimental or approved apps are just installed, nothing special. Official apps must be signed, that is mandatory.

The admin is warned about Integrity problems with a yellow notification bar on top and a link to the documentation.

In the installer I would indeed report integrity problems, point to the documentation and offer to proceed, noting that we strongly advice against that.

I will make sure the upgrade blog, documentation and forums etc contain information about the code integrity checking and how to deal with problems.

**Comment by PVince81** [comment]
- [ ] UX tweaks

@LukasReschke can you fill in the details ? thanks

**Comment by PVince81** [comment]
- [ ] sort out the communication part

**Comment by LukasReschke** [comment]
To merge:

- [ ] https://github.com/owncloud/core/pull/21849
- [ ] https://github.com/owncloud/core/pull/21816
- [ ] https://github.com/owncloud/documentation/pull/2072
- [ ] https://github.com/owncloud/administration/pull/64

**Comment by MorrisJobke** [comment]
@PVince81 @LukasReschke All done - can we close this then?

**Comment by PVince81** [comment]
Are the communication details sorted out also ?

**Comment by LukasReschke** [comment]
> Are the communication details sorted out also ?

Documentation is done as well, needs some small changes by @jospoortvliet though but that's unrelated.

**Comment by PVince81** [comment]
Okay, closing then
